app/services/fact_importer.rb

```diff
@@ -139,17 +137,16 @@ def add_new_facts
   def update_facts
     time = Time.now.utc
     updated = 0
-    db_facts_names = []
-    host.fact_values.includes([:fact_name]).joins(:fact_name).where(FACT_NAME_TYPE => fact_name_class_name).reorder('').find_each do |record|
-      next unless fact_names.include?(record.name)
-      new_value = facts[record.name]
-      if record.value != new_value
+    db_facts = host.fact_values.joins(:fact_name).where(fact_name: {type: fact_name_class_name}).reorder(nil).pluck(:name, :value, :id)
+    db_facts.each do |name, value, id|
+      next unless fact_names.include?(name)
```

Comment:
It feels like this could be captured in a `where` on the query to prevent the DB server from sending it in the first place.

On the other hand, couldn't this data also be used in `delete_removed_facts`? Now we essentially do 2 queries to retrieve the same information. Granted, it's a bigger refactor.

Reply:
Indeed, I tried to keep the logic the same in this PR (and failed.... now pushed with 90% less typos and 60% less logic errors)